iohk/UpdateLogic.hs

```diff
@@ -0,0 +1,344 @@
+  res <- isRepo (FP.decodeString $ T.unpack localpath)
+  case res of
+    True -> do
+      fetchRepo localpath url
+    False -> do
+      exitCode <- proc "git" [ "clone", "--mirror", url, localpath ] mempty
+      case exitCode of
+        ExitSuccess   -> fetchRepo localpath url
+        ExitFailure _ -> error "cant clone repo"
+
+fetchRepo :: T.Text -> T.Text -> IO (Git SHA1)
+fetchRepo localpath url = do
+  let
+    fetcher :: Managed (Git SHA1)
+    fetcher = do
+      pushd  $ FP.decodeString $ T.unpack localpath
+      exitCode <- proc "git" [ "fetch", url ] mempty
+      case exitCode of
+        ExitSuccess -> do
+          liftIO $ putStrLn "fetched"
+          repo <- liftIO $ openRepo $ FP.decodeString $ T.unpack localpath
+          pure repo
+        ExitFailure _ -> error "cant fetch repo"
+  with fetcher $ \res -> pure res
+
+processDarwinBuild :: T.Text -> T.Text -> T.Text -> BuildId -> (T.Text, T.Text) -> T.Text -> IO (GlobalResults, CiResult)
+processDarwinBuild daedalus_rev daedalus_version tempdir buildId (winKey, macosKey) ciUrl = do
+  obj <- fetchTravis buildId
+  let
+    filename = "Daedalus-installer-" <> daedalus_version <> "." <> (number obj) <> ".pkg"
+    version = Version $ daedalus_version <> "." <> (number obj)
+    url = "http://s3.eu-central-1.amazonaws.com/daedalus-travis/" <> filename
+    outFile = tempdir <> "/" <> filename
+  buildLog <- fetchUrl mempty $ "https://api.travis-ci.org/jobs/" <> (T.pack $ show $ tjiId $ head $ drop 1 $ matrix obj) <> "/log"
+  let
+    cardanoBuildNumber = extractBuildId buildLog
+  cardanoInfo <- fetchTravis2 "input-output-hk/cardano-sl" cardanoBuildNumber
+
+  setSGR [ SetColor Background Dull Red ]
+  putStr "cardano build number: "
+  putStrLn $ show cardanoBuildNumber
+
+  putStr "cardano commit: "
+  putStr (T.unpack $ ti2commit cardanoInfo)
+  setSGR [ Reset ]
+  putStr "\n"
+
+  putStr "travis URL: "
+  setSGR [ SetColor Foreground Dull Green ]
+  putStrLn $ T.unpack url
+  setSGR [ Reset ]
+
+  appVersion <- liftIO $ do
+    let
+      readPath name = readFileFromGit (ti2commit cardanoInfo) name "cardano" "https://github.com/input-output-hk/cardano-sl"
+      readPath1 = readPath "lib/configuration.yaml"
+      readPath2 = readPath "node/configuration.yaml"
+      fallback :: GitNotFound -> IO (Maybe LBS.ByteString)
+      fallback _ = readPath2
+    contentMaybe <- catch readPath1 fallback
+    let
+      content = fromMaybe undefined contentMaybe
+      fullConfiguration :: ConfigurationYaml
+      fullConfiguration = fromMaybe undefined (Y.decode $ LBS.toStrict content)
+      winVersion = HashMap.lookup winKey fullConfiguration
+      macosVersion = HashMap.lookup macosKey fullConfiguration
+    -- relies on only parsing out the subset that should match
+    if winVersion == macosVersion then
+      case winVersion of
+        Just val -> do
+          T.putStrLn ("applicationVersion is " <> (T.pack $ show $ applicationVersion $ update val))
```

Comment:
I'm glad this was added. Perhaps as an extra precaution we should require the user to explicitly pass the intended applicationVersion as a parameter and then verify it matches here.

Reply:
i'm thinking that this should generate a json file later, and then the wrapper for propose-update will accept that json file
then it will always propose the same version the yaml claims to be